src/assembly.h

```diff
@@ -341,6 +383,35 @@ static __inline Word64 MADD64(Word64 sum64, int x, int y)
         return u.w64;
 }
 
+/* Ken's trick: clips to [-32768, 32767] */
+//sign = x >> 31;
+//if (sign != (x >> 15))
+//    x = sign ^ ((1 << 15) - 1);
+__attribute__((__always_inline__)) static __inline short SAR64_Clip(Word64 x, int n)
+{
+  unsigned int xLo = (unsigned int) x;
+  int xHi = (int) (x >> 32);
+  int nComp = 32-n;
+  int tmp;
+  // Shortcut: n is always < 32.
+  __asm__ __volatile__( "lsl %2, %0, %3\n\t"  // tmp <- xHi<<(32-n)
+                        "lsr %1, %1, %4\n\t"  // xLo <- xLo>>n
+                        "orr %1, %2\n\t"      // xLo <= xLo || tmp
```

Comment:
I checked what happens if I try to just code SAR64_clip in direct C code and compile it for ARM (gcc 7.mumble).  This code:
```
short SAR64_clip(int64_t x, int n) {
    if(n >= 32) __builtin_unreachable();
    return (x >> n);
}

short SAR64_clip_26(int64_t x) {
    return SAR64_clip(x, 26);
}
```
gets me the sequence
```
00000040 <SAR64_clip_26>:
  40:	0e80      	lsrs	r0, r0, #26
  42:	ea40 1081 	orr.w	r0, r0, r1, lsl #6
  46:	b200      	sxth	r0, r0
  48:	4770      	bx	lr
```
Is this sequence (lsls + orrw-with-lsl) right too?  Is it preferable?  It looks like the SAR64_clip argument is always a compile time constant. (I *assume* that `sxth` is an operation that can be omitted but I'm not 100% sure)

Reply:
The sign extension (sxth) is not needed and doesn't help if the value overflows, but since the value doesn't overflow, we don't have to worry about it. I wasn't sure if you would mind hard coding the shift right value of 26, but it does produce slightly faster code. I just pushed a new version with the simpler code and it's a tiny bit faster.